### PR TITLE
Add new algorithm MatchSpectra

### LINKED ISSUE #21800: Spectrum matching from PDFgetN
In [blend](https://github.com/neutrons/pdfgetn/blob/master/fortran/ublend4.f#L443), there is a section where bank matching is calculated so you add `a*S(Q)+b` where `a` and `b` are calculated from the link above. The algorithm should return the matching and have an option to apply it to the input workspace.

Turns out that this is just an implementation of [Gauss-Markov theorem](https://en.wikipedia.org/wiki/Gauss%E2%80%93Markov_theorem) for a line. However, the implementation in pdfgetn is different than `EstimatePolynomial.cpp`. Porting from FORTRAN will follow the version in `EstimatePolynomial.cpp`.

### PR DESCRIPTION
**Pull request by peterfpeterson**
This is an algorithm that calcuates spectrum matching the same way as PDFgetN.

**Report to:** @marshallmcdonnell 

A code review should be sufficient. Short of using a variety of `S(Q)` spectra, reviewing the unit tests is probably the best way to see that the functionality is being stressed.

Fixes #21800

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/MatchSpectra.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+                sumSpecSq = (Y[tstIndices[0]:tstIndices[1]][mask] * Y[tstIndices[0]:tstIndices[1]][mask]).sum()
+                sumRefSpec = (Y[tstIndices[0]:tstIndices[1]][mask] * referenceY[refIndices[0]:refIndices[1]][mask]).sum()
+
+            # defaults are to do nothing
+            scale = 1.
+            offset = 0.
+
+            if doScale and doOffset:  # use both
+                # Cramar's rule for 2x2 matrix
+                denomenator = totalBins * sumSpecSq - sumSpec * sumSpec
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/MatchSpectra.py]: Should be "denominator"

File: docs/source/algorithms/MatchSpectra-v1.rst [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+Description
+-----------
+
+Calculate the factor(s) that best match the individual spectra to a
+reference spectrum using `Gauss-Markov process
+<https://en.wikipedia.org/wiki/Gauss%E2%80%93Markov_process>`_. This
+is the same algorithm used by the "blend" step of `PDFgetN
+<http://pdfgetn.sourceforge.net/>`_.
+
```
Comment (AndreiSavici) [inline docs/source/algorithms/MatchSpectra-v1.rst]: Should have a formula explaining what you are doing, especially vertical vs horizontal offset.
Also add information about expected failures (such as when all errors are 0)

File: Framework/PythonInterface/plugins/algorithms/MatchSpectra.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                resultScale.append(1.)
+                resultResidual.append(np.nan)
+                continue
+
+            mask = (E[tstIndices[0]:tstIndices[1]] > 0.) * (referenceE[refIndices[0]:refIndices[1]] > 0.)
+            if not np.any(mask):
+                resultOffset.append(0.)
+                resultScale.append(1.)
+                resultResidual.append(np.nan)
+                self.log().information('spectrum {} has no uncertainties greater than zero'.format(spectrumNum))
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/MatchSpectra.py]: Should be warning

File: Framework/PythonInterface/plugins/algorithms/MatchSpectra.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                resultResidual.append(0.)
+                self.log().information('spectrum {} is the reference'.format(spectrumNum))
+                continue
+
+            X = outputWS.readX(wkspIndex)
+            Y = outputWS.readY(wkspIndex)
+            E = outputWS.readE(wkspIndex)
+
+            hasOverlap, refIndices, tstIndices = self.__generateIndices(spectrumNum, referenceX, X, X.size == Y.size+1)
+            if not hasOverlap:
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/MatchSpectra.py]: Add log message

File: Framework/PythonInterface/plugins/algorithms/MatchSpectra.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+                self.log().notice(msg)
+                return BAD_RESULT
+
+        if (reference[refLower:refUpper]).size != (testing[tstLower:tstUpper]).size:
+            self.log().notice(msg)
+            return BAD_RESULT
+
+        return (True, (refLower, refUpper), (tstLower, tstUpper))
+
+    def __residual(self, X, Y1, Y2):
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/MatchSpectra.py]: You are not calculating chi^2 correctly. You need to add errors.

**Review by AndreiSavici — CHANGES_REQUESTED** [review]
You are not using errors (except for masking), therefore your chisq is not really chisq. This should be mentioned in the documentation.

File: Framework/PythonInterface/plugins/algorithms/MatchSpectra.py (hunk 5 above)
Comment (peterfpeterson) [reply Framework/PythonInterface/plugins/algorithms/MatchSpectra.py]: It is the unweighted chi^2 and the documentation has been updated accordingly.

File: docs/source/algorithms/MatchSpectra-v1.rst [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+Calculate the factor(s) that best match the individual spectra to a
+reference spectrum using `Gauss-Markov process
+<https://en.wikipedia.org/wiki/Gauss%E2%80%93Markov_process>`_. This
+is the same algorithm used by the "blend" step of `PDFgetN
+<http://pdfgetn.sourceforge.net/>`_. The data undergoes the linear
+transformation
+
+.. math::
+
+   y_{reference} = scale * y_{spectrum} + offset
```
Comment (AndreiSavici) [inline docs/source/algorithms/MatchSpectra-v1.rst]: should be just `scale * y_{spectrum} + offset` and add "such that the distance between :math:`y_{spectrum}` and the transformed spectrum is minimized."

File: docs/source/algorithms/MatchSpectra-v1.rst (hunk 2 above)
Comment (peterfpeterson) [reply docs/source/algorithms/MatchSpectra-v1.rst]: done

File: Framework/PythonInterface/plugins/algorithms/MatchSpectra.py (hunk 4 above)
Comment (peterfpeterson) [reply Framework/PythonInterface/plugins/algorithms/MatchSpectra.py]: done

**Review by AndreiSavici — APPROVED** [review]

**Pull request merged**